### PR TITLE
Use rsix to make system calls in Wasmtime.

### PR DESCRIPTION
**Pull request by sunfishcode**
`rsix` is a system call wrapper crate that we use in `wasi-common`,
which can provide the following advantages in the rest of Wasmtime:

 - It eliminates some `unsafe` blocks in Wasmtime's code. There's
   still an `unsafe` block in the library, but this way, the `unsafe`
   is factored out and clearly scoped.

 - And, it makes error handling more consistent, factoring out code for
   checking return values and `io::Error::last_os_error()`, and code that
   does `errno::set_errno(0)`.

This doesn't cover *all* system calls; `rsix` doesn't implement
signal-handling APIs, and this doesn't cover calls made through `std` or
crates like `userfaultfd`, `rand`, and `region`.

### CODE REVIEW

**Comment by akirilov-arm** [comment]
@sunfishcode Interesting, in order to fix issue #3310 on AArch64 Linux, we will need to call the `membarrier(MEMBARRIER_CMD_PRIVATE_EXPEDITED_SYNC_CORE, 0)`, and I suppose that would have to go through `rsix` as well - is it implemented in the crate and if not, are there plans to? Currently, the idea is to wait for support in `libc`, but doing it in `rsix` might be a faster route.

**Comment by bjorn3** [comment]
Adding it to rsix would require first adding it to libc I think as rsix falls back to libc when `--cfg rsix_use_libc` is used.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -163,6 +163,7 @@ jobs:
     - run: |
         rustup target add wasm32-unknown-emscripten
         rustup target add armv7-unknown-linux-gnueabihf
+        sudo apt-get install -y gcc-arm-linux-gnueabihf
```
Comment (alexcrichton) [inline .github/workflows/main.yml:166]: How come this was needed with this change? I would have figured that rsix was already part of wasi-common's dep tree to have this required already...

**Review by alexcrichton — APPROVED** [review]

File: .github/workflows/main.yml (hunk 1 above)
Comment (sunfishcode) [reply .github/workflows/main.yml:166]: It's actually an unrelated change; a contributor recently added 32-bit ARM support in rsix, so it now needs cross tools to be cross-compiled.

Comment (alexcrichton) [reply .github/workflows/main.yml:166]: How come this is required at `cargo check`-time though? Did rsix previously not have C code to build and now it does?

Comment (bjorn3) [reply .github/workflows/main.yml:166]: Rsix has two backends. The libc backend and the raw linux syscall backend. The latter requires compiling external assembly and is only used on supported linux targets. Previously 32-bit ARM was not one of such targets and thus would use the libc backend AFAIU.

File: cranelift/native/Cargo.toml [hunk 2]
```diff
@@ -15,7 +15,7 @@ cranelift-codegen = { path = "../codegen", version = "0.77.0", default-features
 target-lexicon = "0.12"
 
 [target.'cfg(target_arch = "s390x")'.dependencies]
-libc = "0.2.95"
+rsix = "0.23.0"
```
[the remark this answers is not in the document]
Comment (sunfishcode) [reply cranelift/native/Cargo.toml:18]: Makes sense. rsix is only used for one thing in Cranelift here, so I'll submit a PR to revert that to just using libc.

[the remark this answers is not in the document]
Comment (bjorn3) [reply cranelift/native/Cargo.toml:18]: Thanks!

**Pull request merged**